2024 Annual Report Hyperledger Indy (#210)
hyperledger/tsc: project-reports/2024/2024-annual-Hyperledger-Indy.md

Compared to 2022, the Pull Request count was down by about 25% to 189 for the year, while the number of contributors was slightly higher, up about 20% to 46. Compared to Q4 2023, the Q1 2024 report changes had the same ratios — down about 25% (to 24) in Pull Requests, up about 20% in contributors (to 13).

While the existing implementation of Indy is extremely stable, and robust — for example, the Sovrin Foundation has been running its networks for 6 years, with 100% uptime since Feb. 2020 — there has been relatively little substantially new functionality added in some time. That may be changing as, in the latter half of 2023, the idea of “Indy Besu” was raised and a successful PoC developed, demonstrating implementing Indy’s public-permissioned network model on top of a Besu network. Progress has already been good in 2024 with the Indy community agreeing on the creation of the [Indy Besu repository](https://github.com/hyperledger/indy-besu) that further moves the implementation forward.

> Indy Besu is a set of independent smart contracts that **can be used for an existing Besu network** or built a new one from scratch. The main idea here is a modular system (modules list below) that allows deploying smart contracts at any time.

Reviewer: Possible addition to clearly identify possible deployment options. I apologize for not immediately noting this point.
```suggestion
Indy Besu is a set of independent smart contracts. The main idea here is a modular system (modules list below) that allows deploying smart contracts at any time for 
* permissioned Indy netwrok (new or migration of exisitng ones)
* existing Besu networks
* existing or new Ethereum permissionless networks
```

Author: Updates made

Typical Indy [Maintainers File](https://github.com/hyperledger/indy-plenum/blob/main/MAINTAINERS.md).

## Project Adoption

Insights from the Hyperledger Indy Ecosystem Summit showed that project adoption is “stable”. It seems those that have adopted Indy continue to stay with it and while there is evidence of new private deployments, we’re not aware of new public ones. There remains little competition for what Indy is today — an easy to deploy, stable, token-less, public, permissioned blockchain that is purpose-built for privacy-preserving verifiable credentials. The best alternative to Indy, and one being investigated as a potential Indy replacement, is Besu, with features that align with the existing Indy concepts.

It should be noted that we continue to see a steady stream of people deploying Indy through questions asked on Discord and in GitHub issues, but we do not have visibility into why they are deploying Indy and if they are continuing to use it.

## Goals

### Performance Against Prior Goals

> As this is the first Hyperledger Indy Annual Report, there is not a 2022 Annual Report to look back on. However, in March 2023, the Indy Contributors group developed this [Indy Roadmap](https://hackmd.io/GeRP00i0Sj-7z4zXn2MB5g?view), an (non-timelined) wish list of technical capabilities we’d like to see in Indy. That roadmap was part of the agenda of the Hyperledger Indy Ecosystem summit, where it was highlighted that in addition to the technical capabilities, a focus was needed on marketing and on influencing regulatory guidance on decentralized identity infrastructure.

Reviewer 2: Is this roadmap URL referenced in any of the Indy repos? I couldn't find it, at least in the READMEs, though I may have missed something.

Author: It is referenced and discussed at the Indy Contributor meetings. I don’t believe it is in any repos — there is not really an umbrella Indy repo — Indy has a set of repos.

From the Wiki: https://wiki.hyperledger.org/display/indy/Indy+Contributors+Meeting

There were several events around which Hyperledger Indy code development efforts were focused over the year:

* Completion of the Ubuntu 20.04 version of Indy Node. Deployers of Indy networks have reported success in deploying the new version.
* The archiving of Ursa, triggering the move of the BLS Signatures library from Ursa to being maintained in the Indy Project.
* The movement of Aries projects to use the “shared components” in place of the Indy SDK. The “shared components” consist of [Aries Askar](https://github.com/hyperledger/aries-askar) (secure storage), [Indy VDR](https://github.com/hyperledger/indy-vdr) (client interface to Indy networks) and [Hyperledger AnonCreds](https://www.hyperledger.org/projects/anoncreds) (AnonCreds implementation).
* The deprecation of the Indy SDK by the end of Q1 2024. Progress has been made on things like closing all open pull requests and closing all issues. We've also removed the Indy SDK from the public (Read the Docs) documentation.

> In addition, as mentioned earlier in the report, significant progress was made in creating an Indy-Besu proof of concept.

Reviewer: Just my thoughts about the possible extension. Feel free to change it or ignore.

```suggestion
In addition, as mentioned earlier in the report, significant progress was made in creating an Indy-Besu proof of concept. Its code was added to the experimental feature branch of Indy Node repository.
However, the development of the project is planned in a new [Hyperledger Indy Besy repository](https://github.com/hyperledger/indy-besu/tree/main).
```

Author: Adjusted.

* Increased awareness of Indy, its usefulness, and a corresponding increase in contributors to enable it to continue to evolve.
* Formally release the Ubuntu 20.04 release of Indy Node.
    * Look into updating the solution to later releases.
* Tools and guidance in updating existing networks to the 20.04 release.
* Continue to support the migration of Indy clients away from the Indy SDK.
    * Update the Indy documentation to point people to current “best practices”.
    * Update and release the Indy VDR component as needed.
* Address the “Corporate Firewall” issue, where users behind a corporate firewall cannot access an Indy network because of the ZMQ ports being used.
* Evolve Indy so as to be ready for when the current implementation will need more functionality, such via the Indy-Besu approach.
* Desired functionality:
    * Ledger Redactibility (aka "tombstoning" transactions)
    * Additional DIDDoc support, such support for additional key types
    * Read replicas for performance
    * Storing other document types on an Indy ledger, such as OCA Bundles
* Progress on the Indy Besu concept, transitioning from the existing PoC and design to a concrete implementation.

>     * Some of the desired functionality above might be done in the Indy Besu implementation.

Reviewer: Just my thoughts about the possible extension. Feel free to change it or ignore.

```suggestion
* Progress on the Indy Besu concept:
    * Transitioning from the existing PoC and design to a concrete implementation
        * Improvement of existing modules
        * Integration with Aries Frameworks
        * Implementation of migration tools
    * Some of the general desired functionality above might be done in the Indy Besu implementation.
```

Author: Added/adjusted.

Compared to 2022, the Pull Request count was down by about 25% to 189 for the year, while the number of contributors was slightly higher, up about 20% to 46. Compared to Q4 2023, the Q1 2024 report changes had the same ratios — down about 25% (to 24) in Pull Requests, up about 20% in contributors (to 13).

> While the existing implementation of Indy is extremely stable, and robust — for example, the Sovrin Foundation has been running its networks for 6 years, with 100% uptime since Feb. 2020 — there has been relatively little substantially new functionality added in some time. That may be changing as, in the latter half of 2023, the idea of “Indy-Besu” was raised and a successful PoC developed, demonstrating implementing Indy’s public-permissioned network model on top of a Besu network. Progress has already been good in 2024 with the Indy community agreeing on the creation of the [Indy-Besu repository](https://github.com/hyperledger/indy-besu) that further moves the implementation forward.

Reviewer 3: Around this migration path, would the intention here be to "migrate" the project over? Or to release an "indy-besu" version of the indy project? Re-using the same repo has pros and cons, like forks, stars, contributor base, and more. Or would a new project be opened under Hyperledger? There is another view where this could be added to the Besu project as a "plug-in" or even a consensus model. Curious on thoughts here.

Author: The community has decided to create an [indy-besu repo](https://github.com/hyperledger/indy-besu) and to create an implementation. There is a designed migration path of an existing indy-node to an indy-besu deployment.

The PoC was put into the same repo, but the work will be in the new repo going forward.

Reviewer: @non-fungible-nelson Let me just give a brief background of this decision:
* Hyperledger Indy Ecosystem Summit – June 29, 2023 - Community discussed Indy issues and possibility of implementing a "new" Indy.
* Autumn Indy Summit - DSR presented the idea and options of implementing the new Indy. After confirmation of the community's interest, implementation of the PoC has been begun.
* The end of 2023 - PoC is published the Indy Node repo, as proof of possibility of using Hyperledger Besu for Indy business logic - https://github.com/hyperledger/indy-node/tree/indy-besu
* Beginning 2024 - The implementation of MVP in a new repository in Hyperledger Indy continues, as it includes a large amount of code: smart contracts + VDR + documentation.

The Hyperledger Indy ecosystem community came together at a series of Hyperledger Indy Ecosystem Summits (organized by the Sovrin Foundation), bringing together network operators, node operators, project maintainers, network users, vendors to discuss the state of Indy and proposals for moving the project forward. The summits were well-attended (up to 90 participants), a better understanding of the community of interest was gained, and next steps were proposed, with some (mostly notably, the Indy-Besu initiative) acted upon.

## Maintainer Diversity

_Guidance: How many maintainers do you have, and which organisations are they from? How has the maintainers and diversity of your maintainers changed in the past year? Has the number of active maintainers increased/decreased? Has the diversity of maintainers increased/decreased? Please include a link to your existing [MAINTAINERS file](https://github.com/hyperledger/toc/blob/gh-pages/guidelines/MAINTAINERS-guidelines.md) and the MAINTAINERS file from last year (if appropriate). This is a good opportunity to ensure that your MAINTAINERS file is up to date and to retire any maintainers._

> There are 35 individuals on 22 Indy GitHub Teams representing at least 16 organizations. We continue to have participation from (more or less) the same organizations as in 2022 — which is both good (still participating) and bad (few newcomers).

Reviewer 3: Does Indy-Besu provide any synergy here? Could see benefit in providing OOTB solution for Besu networks that already exist too - could grow contributors on both sides. I.e. add identity to any Besu net could attract a lot of devs using Besu...

**Tangent zone...** and make the existing  permissioning features obsolete? Very curious about synergies in the in-built Besu features. There is the Indy-Besu using the underlying ledger, and Identity providing novel features to Besu ledgers/networks themselves. Would love to explore this space.

Author: I think there is definitely the possiblity of synergy with Besu. A couple of ways of doing this:

- Simply enable putting DIDs and AnonCreds objects on any Besu networks. Work is happening in the community to do that — [AnonCreds on Ethereum](https://github.com/gmulhearn/anoncreds-on-ethereum).  That is the “permission-less” approach of just publishing objects for identity in a public place.
- The idea of indy-besu is definitely for an OOTB deployment of Besu that is enabled for Indy use-cases.  I’m not sure about adding to an existing Besu deployment — a good question for those that have done the Indy Besu design.

Reviewer: One of the main ideas behind Indy Besu was the goal of creating synergy between Hyperledger Indy and Besu to strengthen and extend both communities.

Indy Besu is a set of independent smart contracts that **can be used for an existing Besu network** or built a new one from scratch.
The main idea here is a modular system that allows to deploy smart contracts at any time
### Modules

- Network Permission modules:
  - [Auth](https://github.com/hyperledger/indy-besu/blob/main/docs/design/auth.md) - control user permissions
    - role control - manage roles assigned to accounts  
    - access control - first level validation: whether to accept write transactions (execute target contract method) from a given account
  - [Upgrading contracts](https://github.com/hyperledger/indy-besu/blob/main/docs/design/upgradability.md) - control versions of deployed contracts (proposing and approving new versions).
  - [Validators node management](design/network.md) - control the list of network validator nodes
- Identity: 
  - [DID Method](https://github.com/hyperledger/indy-besu/blob/main/docs/design/did-method.md) - Supported DID method
  - [CL Registry](https://github.com/hyperledger/indy-besu/blob/main/docs/design/cl-registry.md)
- [Client](https://github.com/hyperledger/indy-besu/blob/main/docs/design/vdr.md) - design of VDR library

More links and docs here https://github.com/hyperledger/indy-besu/blob/main/docs/README.md

Reviewer: One of the main ideas behind Indy Besu was the goal of creating synergy between Hyperledger Indy and Besu to strengthen and extend both communities.

Indy Besu is a set of independent smart contracts that **can be used for an existing Besu network** or built a new one from scratch.
The main idea here is a modular system that allows to deploy smart contracts at any time
### Modules

- Network Permission modules:
  - [Auth](https://github.com/hyperledger/indy-besu/blob/main/docs/design/auth.md) - control user permissions
    - role control - manage roles assigned to accounts  
    - access control - first level validation: whether to accept write transactions (execute target contract method) from a given account
  - [Upgrading contracts](https://github.com/hyperledger/indy-besu/blob/main/docs/design/upgradability.md) - control versions of deployed contracts (proposing and approving new versions).
  - [Validators node management](design/network.md) - control the list of network validator nodes
- Identity: 
  - [DID Method](https://github.com/hyperledger/indy-besu/blob/main/docs/design/did-method.md) - Supported DID method
  - [CL Registry](https://github.com/hyperledger/indy-besu/blob/main/docs/design/cl-registry.md)
- [Client](https://github.com/hyperledger/indy-besu/blob/main/docs/design/vdr.md) - design of VDR library

More links and docs here https://github.com/hyperledger/indy-besu/blob/main/docs/README.md

Author: Added this content to the section about Indy Besu.

From the last Quarterly Report: _Still needed -- a cleanup of the teams are needed as some of the individuals are no longer active in the community. No progress was made on that cleanup this quarter. We have the team lists, but have not queried those on the list to see if they are still interested in being maintainers._

> Typical Indy [Maintainers File](https://github.com/hyperledger/indy-plenum/blob/main/MAINTAINERS.md).

Reviewer 2: Several maintainers' entries in the "Active Maintainers" table are incomplete, with no emails, Discord handles, and affiliations. Ideally, you should display this info so other users and potential contributors can reach out to them. I do see some of those maintainers in the Indy discord channels, so I don't believe there's a question of privacy here.

## Project Adoption

_Guidance: What do you know about adoption, and how has this changed since your last review or since being accepted into Hyperledger Foundation? If you can list companies that are adopters of your project, please do so. Feel free to link to an existing ADOPTERS file if appropriate._

> Insights from the Hyperledger Indy Ecosystem Summit showed that project adoption is “stable”. It seems those that have adopted Indy continue to stay with it and while there is evidence of new private deployments, we’re not aware of new public ones. There remains little competition for what Indy is today — an easy to deploy, stable, token-less, public, permissioned blockchain that is purpose-built for privacy-preserving verifiable credentials. The best alternative to Indy, and one being investigated as a potential Indy replacement, is Besu, with features that align with the existing Indy concepts.

Reviewer 3: See above comment - these can definitely be linked in a way to boost both projects.

Reviewer 2: Just to add to what this report says and what Matt has noted: there's an ongoing feature being developed (in a mentorship project) within core Cacti that necessitates a dependency to Indy.

Two questions related to this (putting on my hat as a mentor as well as a reviewer):
- As you have highlighted above, there is relatively little competition for Indy in its niche, and that is what a lot of adopters and users find compelling about it. But oftentimes those users may need additional pluggable features (as core Indy is quite generic). Are there established guidelines and procedures for people wishing to add such features to Indy?
- Since the big refactoring of Indy and Aries, there seems to be relatively little guideline for people who were familiar with the older repositories (including `indy-sdk`) to migrate to the new ones. Is Discord the only way to get such guidance or are there more "static" references people can use?

Author: * There are not much in the way of guidelines/procedures for those wanting to add features to Indy. The community is there to help with questions and guidance, but we do not have nearly enough proactive guidance.
* For those that used the Aries Frameworks, there is guidance for moving a deployment from the Indy SDK to [Aries Askar](https://github.com/hyperledger/aries-askar) and the other components (such as [this](https://aca-py.org/latest/deploying/IndySDKtoAskarMigration/) for Aries Cloud Agent Python). However, there is not much guidance for those that have built their own interfaces based on the indy-sdk.

As this is the first Hyperledger Indy Annual Report, there is not a 2022 Annual Report to look back on. However, in March 2023, the Indy Contributors group developed this [Indy Roadmap](https://hackmd.io/GeRP00i0Sj-7z4zXn2MB5g?view), an (non-timelined) wish list of technical capabilities we’d like to see in Indy. That roadmap was part of the agenda of the Hyperledger Indy Ecosystem summit, where it was highlighted that in addition to the technical capabilities, a focus was needed on marketing and on influencing regulatory guidance on decentralized identity infrastructure.

Of those items, progress was made on the following items:

* Ubuntu 20.04 development, deployment, upgrade documentation, and release.
* Clean up of issues in the various Indy repositories.
* Indy node monitoring capabilities.

> That is less progress than we would have liked to see. A positive spin on that lack of progress is that Indy continues to be stable and adequately scalable such that there is no need 

Reviewer 2: It's clear what you are trying to say, but it looks like you are missing a few words at the end of the sentence.

Author: Updated to finish the sentence 😄

There were several events around which Hyperledger Indy code development efforts were focused over the year:

* Completion of the Ubuntu 20.04 version of Indy Node. Deployers of Indy networks have reported success in deploying the new version.
* The archiving of Ursa, triggering the move of the BLS Signatures library from Ursa to being maintained in the Indy Project.
* The movement of Aries projects to use the “shared components” in place of the Indy SDK. The “shared components” consist of [Aries Askar](https://github.com/hyperledger/aries-askar) (secure storage), [Indy VDR](https://github.com/hyperledger/indy-vdr) (client interface to Indy networks) and [Hyperledger AnonCreds](https://www.hyperledger.org/projects/anoncreds) (AnonCreds implementation).

> * The deprecation of the Indy SDK.

Reviewer 2: Is there a timeline to move from deprecation to archival? (I'm sure there is in the roadmap, but since you mention it here, a tentative date would be good to add.)

Author: Added a goal of end of Q1 2024. Good point — thanks.

> In addition, as mentioned earlier in the report, significant progress was made in creating an Indy-Besu proof of concept.

Reviewer 2: Can you say a bit more about the extent of coupling between Indy and Besu in this PoC? For example, since Besu comes with its own pluggable consensus protocol (e.g., IBFT, QBFT), does Indy-Besu involve a clean replacement of the Indy Plenum with a Besu consensus module?

I'm not trying to get technical details, but just trying to understand what this tells us about Indy's future evolution and what lessons others can take. For example, if one wants to create an Indy-Fabric, how would they envision it?

Author: I’m guessing you are asking out of interest and not to be included in the Indy Annual report, so I’ve not added this to the report.

The idea is to bring up a new network with the roles/permissions capabilities of Indy added to Besu, to enable the same concepts to exist — a Trustee, Steward, Endorser, Author, and a configuration of permissions about what is needed from each to execute a transaction, and a comparable SDK model.  The implementation can be extended to enable migrating transactions from an existing Indy Node deployment such that the old ledger can be shutdown, with its transactions still resolvable.

Reviewer 2: Right, I didn't mean for this to be included in the annual report. It does make me sanguine about the project's health and vision, and thereby informs my review, though. Thanks!

### Next Year’s Goals

_Guidance: What are the goals for the next year of the project? The goals should list what you want to achieve, not just what you know you can achieve. Feel free to include stretch goals and things that you are looking to explore in the next year. For example, are you working on major new features? Or are you concentrating on adoption, community growth, or documentation?_

> The goals for 2024 derive from both 2023 Indy Roadmap document, and the outcome of the Indy Ecosystems Summits held in 2023. The bullet point versions of these goals include the following. The Indy maintainers and contributors have a deeper understanding of what is behind these bullet points:

Reviewer 3: Indy-Besu goals? Or are we considering a separate work bucket?

Author: Doh—good point, we should have had it in there. Added and noted that some of the goals might be accomplished via Indy Node or in Indy Besu.

* Increased awareness of Indy, its usefulness, and a corresponding increase in contributors to enable it to continue to evolve.
* Formally release the Ubuntu 20.04 release of Indy Node.
    * Look into updating the solution to later releases.
* Tools and guidance in updating existing networks to the 20.04 release.
* Continue to support the migration of Indy clients away from the Indy SDK.
    * Update the Indy documentation to point people to current “best practices”.
    * Update and release the Indy VDR component as needed.
* Address the “Corporate Firewall” issue, where users behind a corporate firewall cannot access an Indy network because of the ZMQ ports being used.

> * Evolve Indy so as to be ready for when the current implementation will need more functionality, such via the Indy-Besu approach.

Reviewer 2: Thanks for mentioning this. This relates to my previous comment. Can you provide any specifics about the nature of this evolution? Is it about making the system more modular and pluggable? (I believe your refactoring was intended to serve that purpose.)

Author: Here are a summary of the main concerns:

- The ease of adding more types of transactions so that new object types can be published on the ledger with some verification. Currently, that requires more work than we would like, so it does not happen as often as we would need.
- With more types of transactions we can more easily transition to new of kinds of verifiable credentials, such as AnonCreds v2.
- With transitions to new VC formats, we get better functionality, such as much better revocation — which will reduce the number of transactions on the ledger — reducing the pressure on scalability.
- Scalability is not an issue today, but will be some day.
- It would be much better to be part of a bigger community.

But with all those concerns — none of them are serious enough to get people refocused from building on Indy to evolving Indy.

Reviewer 2: Thanks @swcurran . This answers my question (and satisfies my curiosity).

It's not mandatory, but if you wish, you can add these bullet points to the report too, at the above location. I think the community would benefit from your thoughts on the future evolution of the project.

Author: Added those items.  All of the goals are subject to completion in either or both of the Indy or Indy Besu implementations.

* Desired functionality:
    * Tombstoning
    * Additional DIDDoc support, such support for additional key types
    * Read replicas for performance
    * Storing other document types on an Indy ledger, such as OCA Bundles

### Help Required

_Guidance: How can the Hyperledger Foundation or the TOC help you achieve your upcoming goals?_

> No specific requests at this time. There is a lot that the Hyperledger Foundation provides that we need to be making use of, particularly in the area of marketing / getting the word out about the capabilities of Hyperledger Indy. Our challenge is that the project is driven largely by technical types, not marketers.

Reviewer 2: Other Hyperledger projects typically conduct periodic workshops and write Hyperledger blog articles for this purpose. @davidwboswell can you advise the maintainers and help them with promotion and advocacy?

Author: @davidwboswell has been a huge help in this area. The hard part is just getting the time from contributors to provide the content to take advantage of all that Hyperledger offers.

## Project Lifecycle Stage Recommendation

_Guidance: What stage do you think the project should be? If you think that your project meets the criteria for another stage, please explain why. (See [Project Lifecycle](https://github.com/hyperledger/toc/blob/gh-pages/governing-documents/project-lifecycle.md))_

> Indy is a Graduated project and should remain so. Its code is rock solid, and a relatively large ecosystem of network operators, maintainers, contributors, and users.

Reviewer 3: Overall, I think this is great and tracks with the project maintaining its status. Apologies in advance for harping on Indy-Besu :)

Reviewer 2: +1 to what Matt said.
